### PR TITLE
[RefundPayment] Make the Refund customization easier

### PR DESCRIPTION
**Pull request by GSadee**

### CODE REVIEW

File: src/Factory/RefundPaymentFactory.php [hunk 1]
```diff
@@ -15,30 +15,25 @@
...
-    private $paymentMethodRepository;
+    /** @var string */
+    private $className;
 
-    public function __construct(PaymentMethodRepositoryInterface $paymentMethodRepository)
+    public function __construct(string $className)
     {
-        $this->paymentMethodRepository = $paymentMethodRepository;
+        $this->className = $className;
     }
```
Comment (lchrusciel) [inline src/Factory/RefundPaymentFactory.php:28]: This change is not announced in changelog

**Review by lchrusciel — APPROVED** [review]

File: src/Factory/RefundPaymentFactory.php (hunk 1 above)
Comment (GSadee) [reply src/Factory/RefundPaymentFactory.php:28]: Added 👌🏻

File: spec/Creator/RefundUnitsCommandCreatorSpec.php [hunk 2]
```diff
@@ -25,21 +25,24 @@
 
 final class RefundUnitsCommandCreatorSpec extends ObjectBehavior
 {
-    public function let(
-        UnitRefundTotalCalculatorInterface $unitRefundTotalCalculator
-    ): void {
-        $this->beConstructedWith($unitRefundTotalCalculator);
+    public function let(RefundUnitsConverterInterface $refundUnitsConverter): void
```
Comment (Zales0123) [inline spec/Creator/RefundUnitsCommandCreatorSpec.php:28]: `public` in spec 💃

File: spec/Factory/RefundPaymentFactorySpec.php [hunk 3]
```diff
@@ -39,29 +38,45 @@ public function it_implements_refund_payment_factory_interface(): void
...
-            ->shouldBeLike(
-                new RefundPayment(
-                    $order->getWrappedObject(),
-                    1000,
-                    'USD',
-                    RefundPaymentInterface::STATE_NEW,
-                    $paymentMethod->getWrappedObject()
-                )
-            );
+            ->shouldBeSameAs($refundPayment)
```
Comment (Zales0123) [inline spec/Factory/RefundPaymentFactorySpec.php:60]: What was wrong with `shouldBeLike`? With this approach, we need to mock the expected result 🚀

File: spec/ProcessManager/RefundPaymentProcessManagerSpec.php [hunk 4]
```diff
@@ -59,20 +63,26 @@ public function it_reacts_on_units_refunded_event_and_creates_refund_payment(
...
-            1
-        )->willReturn($refundPayment);
+        $refundPaymentFactory
+            ->createWithData(
+                $order,
+                1000,
+                'USD',
+                RefundPaymentInterface::STATE_NEW,
+                $paymentMethod
+            )
```
Comment (Zales0123) [inline spec/ProcessManager/RefundPaymentProcessManagerSpec.php:83]: Wouldn't it fit in one line? 😄

File: src/Converter/RefundUnitsConverter.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public function convert(array $units, RefundType $refundType, string $unitRefundClass): array
+    {
+        $units = $this->filterEmptyRefundUnits($units);
+        $refundUnits = [];
+        foreach ($units as $id => $unit) {
+            $total = $this->unitRefundTotalCalculator
+                ->calculateForUnitWithIdAndType($id, $refundType, $this->getAmount($unit))
+            ;
```
Comment (Zales0123) [inline src/Converter/RefundUnitsConverter.php:37]:
```php
$total = $this
    ->unitRefundTotalCalculator
    ->calculateForUnitWithIdAndType($id, $refundType, $this->getAmount($unit))
;
```

IMO

File: src/Converter/RefundUnitsConverter.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        return array_filter($units, function (array $refundUnit): bool {
+            return (isset($refundUnit['amount']) && $refundUnit['amount'] !== '') || isset($refundUnit['full']);
+        });
+    }
+
+    private function getAmount(array $unit): ?float
+    {
+        if (isset($unit['full'])) {
+            return null;
+        }
```
Comment (Zales0123) [inline src/Converter/RefundUnitsConverter.php:56]: This is so bad :( I wonder, cannot we just always operate on the actual value? I don't know won't it need a avalanche of changes in other parts of the application, so could possibly be done as an improvement in a separate PR

**Review by Zales0123 — APPROVED** [review]

File: spec/Creator/RefundUnitsCommandCreatorSpec.php (hunk 2 above)
Comment (GSadee) [reply spec/Creator/RefundUnitsCommandCreatorSpec.php:28]: Why not? `public` is a default one if there is not access modifier specified

File: spec/Factory/RefundPaymentFactorySpec.php (hunk 3 above)
Comment (GSadee) [reply spec/Factory/RefundPaymentFactorySpec.php:60]: To be honest, I've used the exisiting comparator

File: src/Converter/RefundUnitsConverter.php (hunk 6 above)
Comment (GSadee) [reply src/Converter/RefundUnitsConverter.php:56]: I agree, but I didn't want to change the current behaviour, just extract some logic and improve extendability

**Comment by Zales0123** [comment]
Thank you, Grzegorz! :tada:
